File: RxDiRetrofitApp/japp/src/main/java/com/woowrale/rxdiretrofit/ui/main/MainPresenter.java
package com.woowrale.rxdiretrofit.ui.main;

import com.woowrale.rxdiretrofit.data.model.Crypto;
import com.woowrale.rxdiretrofit.data.repository.ApiService;
import com.woowrale.rxdiretrofit.ui.base.BasePresenter;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class MainPresenter extends BasePresenter<MainView> {

    private CompositeDisposable compositeDisposable;

    public void onSearchCryptoCurrency(){

        compositeDisposable = new CompositeDisposable();
        compositeDisposable.add(ApiService.services().getCoinData("btc")
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(this::handleResults));
    }


    private void handleResults(Crypto crypto) {
        if (crypto != null) {
            if(crypto.ticker.markets != null){
                getView().showListView(crypto.ticker.markets);
            }
        } else {
            getView().showMessageResult();
        }
    }

    private void handleError(Throwable t) {
        getView().showMessageError();
    }

}

File: RxDiRetrofitApp/kapp/src/main/java/com/woowrale/rxdiretrofit/ui/main/MainPresenter.kt
package com.woowrale.rxdiretrofit.ui.main

import android.annotation.SuppressLint
import android.util.Log
import com.woowrale.rxdiretrofit.data.model.Crypto
import com.woowrale.rxdiretrofit.data.repository.ApiServices
import com.woowrale.rxdiretrofit.ui.base.BasePresenter

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class MainPresenter : BasePresenter<MainView>() {

    val TAG_NAME: String = this.javaClass.canonicalName
    var compositeDisposable: CompositeDisposable? = null

    @SuppressLint("CheckResult")
    fun onCallEndPoint() {
        compositeDisposable = CompositeDisposable()
        compositeDisposable?.add(ApiServices.service().getCoinData("btc")
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe({t -> handleResults(t)}))
    }

    private fun handleResults(crypto: Crypto) {

        if (crypto != null) {
             getView()?.setRecyclerViewAdapterData(crypto.ticker?.markets!!)
        } else {
            getView()?.showMessageResult()
        }
    }

    private fun handleError(t: Throwable) {
        Log.e(TAG_NAME, t.message)
        getView()?.showMessageError()
    }
}